tutorials/src/bin/basic-tutorial-9.rs

```diff
@@ -0,0 +1,184 @@
+fn print_topology(info: &DiscovererStreamInfo, depth: usize) {
+    print_stream_info(info, depth);
+
+    if let Some(next) = info.get_next() {
+        print_topology(&next, depth + 1);
+    } else if let Some(container_info) = info.downcast_ref::<DiscovererContainerInfo>() {
+        for stream in container_info.get_streams() {
+            print_topology(&stream, depth + 1);
+        }
+    }
+}
+
+fn on_connect_discovered(
+    _discoverer: &Discoverer,
+    discoverer_info: &DiscovererInfo,
+    error: &Option<glib::Error>,
+) {
+    let uri = discoverer_info.get_uri().unwrap();
+    match discoverer_info.get_result() {
+        DiscovererResult::Ok => println!("Discovered {}", uri),
+        DiscovererResult::UriInvalid => println!("Invalid uri {}", uri),
+        DiscovererResult::Error => if let Some(msg) = error {
+            println!("{}", msg);
+        } else {
+            println!("Unknown error")
+        },
+        DiscovererResult::Timeout => println!("Timeout"),
+        DiscovererResult::Busy => println!("Busy"),
+        DiscovererResult::MissingPlugins => if let Some(s) = discoverer_info.get_misc() {
+            println!("{}", s);
+        },
+        _ => println!("Unknown result"),
+    }
+
+    if discoverer_info.get_result() != DiscovererResult::Ok {
+        return;
+    }
+
+    println!("Duration: {}", discoverer_info.get_duration());
+
+    if let Some(tags) = discoverer_info.get_tags() {
+        println!("Tags:");
+        for (tag, values) in tags.iter_generic() {
+            print!("  {}: ", tag);
+            values.for_each(|v| {
+                if let Some(s) = send_value_as_str(v) {
+                    println!("{}", s)
+                }
+            })
+        }
+    }
+
+    println!(
+        "Seekable: {}",
+        if discoverer_info.get_seekable() {
+            "yes"
+        } else {
+            "no"
+        }
+    );
+
+    println!("Stream information:");
+
+    if let Some(stream_info) = discoverer_info.get_stream_info() {
+        print_topology(&stream_info, 1);
+    }
+}
+
+fn run_discoverer() -> Result<(), Error> {
+    gst::init()?;
+
+    let args: Vec<_> = env::args().collect();
+    let uri: &str = if args.len() == 2 {
+        args[1].as_ref()
+    } else {
+        "https://www.freedesktop.org/software/gstreamer-sdk/data/media/sintel_trailer-480p.webm"
+    };
+
+    println!("Discovering {}", uri);
+
+    let loop_ = glib::MainLoop::new(None, false);
+    let timeout = gst::ClockTime::from_seconds(5);
+    let discoverer = gst_pbutils::Discoverer::new(timeout)?;
+    discoverer.connect_discovered(on_connect_discovered);
+    let loop_clone = loop_.clone();
+    discoverer.connect_finished(move |_| {
+        println!("\nFinished discovering");
+        loop_clone.quit();
+    });
+    discoverer.start();
+    discoverer.discover_uri_async(uri)?;
+
+    loop_.run();
+
+    discoverer.stop();
+
+    Ok(())
+}
+
+fn tutorial_main() {
```

Comment:
You could make this return a `Result<(), Error>`, and the same for `main()`

Reply:
I needed some indirection because having the trailing return type on main was making the compiler upset, since it appears `tutorials_common::run` doesn't like the return value.